### ISSUE TITLE
Regression with constrained TypeVars (v1.1.315 -> v1.1.316)

### ISSUE THREAD

**Issue by runfalk** [microsoft/pyright#5456]
**Describe the bug**
When using constrained `TypeVar`s in a dataclass, the type checking works differently depending on whether the constructor is called through a class method or its name.

**To Reproduce**
The following program exhibits the problematic behavior:

```python
import dataclasses
import typing as t

T_value = t.TypeVar(
    "T_value",
    str,  # Removing this makes type checking pass
    int,
    float,
)

@dataclasses.dataclass(frozen=True)
class KeyValue(t.Generic[T_value]):
    key: str
    value: T_value | None

    @classmethod
    def default(cls) -> "KeyValue[t.Any]":
        # This doesn't work
        return cls("foobar", 1.1)  # Works when using KeyValue instead of cls

# These do work
def factory() -> KeyValue[t.Any]:
    return KeyValue("foobar", 1.1)

kv: KeyValue[t.Any] = KeyValue("foobar", 1.1)
KeyValue("foobar", 1.1)
```

**Expected behavior**
I expect the entire program to pass type checking.

**Actual behavior**
```
/x.py
  /x.py:19:30 - error: Argument of type "float" cannot be assigned to parameter "value" of type "T_value@KeyValue | None" in function "__init__"
    Type "float" cannot be assigned to type "T_value@KeyValue | None"
      Type "float" is incompatible with constrained type variable "T_value@KeyValue"
      Type cannot be assigned to type "None" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Additional context**
Interestingly removing the `str` constraint actually "fixes" the problem and allows type checking to pass.

The regression seems to have been introduced in `1.1.316` since `1.1.315` passes for the same example program.

**Comment by erictraut** [comment]
Pyright is correct in generating an error here. The type of `cls` is already specialized in a class method. Its type is `type[Self[T_value]]`, which means that if you call `cls` as a constructor, `T_value` must be consistent. You're passing a value of type `float` to the constructor when you must pass a value of type `T_value`.

This previously did not generate a type error in pyright because of a bug which I recently fixed. Mypy correct generates a type error here.

**Comment by runfalk** [comment]
What you're saying makes sense, but why does removing `str` from the constraints allow type checking to pass?

Does this change mean it's impossible to refer to the `cls` constructor in generic classes? Since how can I get a type that is `T_value`?

**Comment by erictraut** [comment]
If you remove the `str` from the constraints, it effectively collapses to a `float` because `int` is a subtype of `float`. This is one of the (many) oddities of constrained TypeVars.

Are you sure you should be using a constrained TypeVar for your use case? About 90% of the uses of constrained TypeVars that I see are inappropriate. It's unfortunate that constrained TypeVars were added to the Python type system, but it's too late to remove them now. To my knowledge, no other type language has this odd construct — and for good reason.

> Does this change mean it's impossible to refer to the cls constructor in generic classes?

You can construct an instance of the same class (one with the same type argument `T_value`), but you can't construct an instance of the class with a different type argument. To do so, you would need to use a `cast`.

**Comment by runfalk** [comment]
Thank you for the clear explanation!

> If you remove the str from the constraints, it effectively collapses to a float because int is a subtype of float

Right! I forgot about the `float -> int -> bool` relationship. Some class hierarchies are a bit annoying to work with in Python.

> Are you sure you should be using a constrained TypeVar for your use case?

I think the use case is fairly reasonable. We are deserializing a list of key value tuples. There are a limited number of different types that we can deserialize into and we identify them using an enum. For every entry in the enum there is a corresponding constraint in the typevar. AFAIK the constrainted typevars are the only way to precisely control this as `bound=` would allow subclasses as well?

We are also using it for the same reason in an ORM style framework to make it impossible to set up attributes on the mapper objects that can't be safely serialized into the DB.
